File: src/main/java/org/evernet/bean/NodeAddress.java
package org.evernet.bean;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.*;

import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(NON_NULL)
@EqualsAndHashCode
public class NodeAddress {

    private String identifier;

    private String vertexEndpoint;

    public static NodeAddress fromString(String s) {
        String[] components = s.split("/");

        if (components.length != 2) {
            throw new IllegalArgumentException(String.format("Invalid node address %s", s));
        }

        return NodeAddress.builder().vertexEndpoint(components[0]).identifier(components[1]).build();
    }

    public String toString() {
        return String.format("%s/%s", vertexEndpoint, identifier);
    }
}

File: internal/pkg/address/node_address.go
package address

import (
	"errors"
	"fmt"
	"strings"
)

type NodeAddress struct {
	Identifier     string `json:"identifier"`
	VertexEndpoint string `json:"vertex_endpoint"`
}

func (n *NodeAddress) ToString() string {
	return fmt.Sprintf("%s/%s", n.VertexEndpoint, n.Identifier)
}

func NodeAddressFromString(nodeAddress string) (*NodeAddress, error) {
	components := strings.Split(nodeAddress, "/")

	if len(components) != 2 {
		return nil, errors.New("invalid node address")
	}

	identifier := components[1]
	vertexEndpoint := components[0]

	return &NodeAddress{
		Identifier:     identifier,
		VertexEndpoint: vertexEndpoint,
	}, nil
}
